File: integration-test/java/jee/src/main/java/com/qualityminds/lazyval/integration/boundary/rest/RequestLocaleMessageInterpolator.java
package com.qualityminds.lazyval.integration.boundary.rest;

import jakarta.validation.MessageInterpolator;
import jakarta.validation.Validation;

import java.util.Locale;

/**
 * Wraps the provider's default message interpolator and re-interprets the two-argument
 * {@link #interpolate(String, Context)} call so it uses the per-request locale stored in
 * {@link RequestLocaleHolder} instead of {@code Locale.getDefault()}.
 *
 * Registered as {@code <message-interpolator>} in {@code META-INF/validation.xml}, hence
 * the public no-arg constructor.
 */
public class RequestLocaleMessageInterpolator implements MessageInterpolator {

    private final MessageInterpolator delegate;

    public RequestLocaleMessageInterpolator() {
        this.delegate = Validation.byDefaultProvider().configure().getDefaultMessageInterpolator();
    }

    @Override
    public String interpolate(String messageTemplate, Context context) {
        Locale locale = RequestLocaleHolder.get();
        return locale != null
                ? delegate.interpolate(messageTemplate, context, locale)
                : delegate.interpolate(messageTemplate, context);
    }

    @Override
    public String interpolate(String messageTemplate, Context context, Locale locale) {
        return delegate.interpolate(messageTemplate, context, locale);
    }
}

File: integration-test/kotlin/jee/src/main/kotlin/com/qualityminds/lazyval/integration/boundary/rest/RequestLocaleMessageInterpolator.kt
package com.qualityminds.lazyval.integration.boundary.rest

import jakarta.validation.MessageInterpolator
import jakarta.validation.Validation
import java.util.Locale

/**
 * Wraps the provider's default message interpolator and re-interprets the two-argument
 * [interpolate] call so it uses the per-request locale stored in [RequestLocaleHolder]
 * instead of `Locale.getDefault()`.
 *
 * Registered as `<message-interpolator>` in `META-INF/validation.xml`, hence the public
 * no-arg constructor.
 */
class RequestLocaleMessageInterpolator : MessageInterpolator {

    private val delegate: MessageInterpolator =
        Validation.byDefaultProvider().configure().defaultMessageInterpolator

    override fun interpolate(messageTemplate: String, context: MessageInterpolator.Context): String {
        val locale = RequestLocaleHolder.get()
        return if (locale != null) {
            delegate.interpolate(messageTemplate, context, locale)
        } else {
            delegate.interpolate(messageTemplate, context)
        }
    }

    override fun interpolate(
        messageTemplate: String,
        context: MessageInterpolator.Context,
        locale: Locale,
    ): String = delegate.interpolate(messageTemplate, context, locale)
}
